_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,429 @@
+- [Limitations, Assumptions & Future Work](#limitations-assumptions--future-work)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Installation
+
+If you want to follow along with the examples in this blog post, you need to install the
+[`nix`](https://nixos.org/) package manager, and then install the
+[`dapptools`](https://github.com/dapphub/dapptools) toolkit:
+
+```sh
+# user must be in sudoers
+curl -L https://nixos.org/nix/install | sh
+
+# Run this or login again to use Nix
+. "$HOME/.nix-profile/etc/profile.d/nix.sh"
+
+# install dapptools
+curl https://dapp.tools/install | sh
+```
+
+All the examples from this blog post are available [here](https://github.com/xwvvvvwx/ds-test-blog-examples).
+
+## What Is Symbolic Execution?
+
+Symbolic execution is a program analysis technique that keeps some of the program state in an
+abstract form, meaning that instead of being set to a specific value, these portions of the state
+are represented as a name with some constraints attached.
+
+To make this more specific, consider the contract below:
+
+```solidity
+contract Add {
+    function add(uint x, uint y) external pure returns (uint z) {
+        require((z = x + y) >= x, "overflow!");
+    }
+}
+```
+
+If we were to execute the `add` method symbolically we could represent the calldata as two symbolic
+variables (`x` and `y`) that are constrained so their value can only fit in the range of a
+`uint256`. As we proceed through the program, we will encounter various potential branching points,
+for example a `JUMPI` instruction. If both sides of the branching point are reachable (determined by
+checking if the conjunction of all existing constraints and the jump condition is satisfiable), then
+execution will split in two, and each branch will be explored seperately, with the jump condition
+(or it's negation depending on which branch is being explored) being added as a constraint for that
+particular branch.
+
+This results in a tree of possible executions, for example for the `add` method, the execution tree
+looks like this:
+
+```
+├ 0     msg.value > 0
+│       Revert
+│
+└ 1     msg.value == 0
+  │
+  ├ 0   x + y < x
+  │     Revert("overflow!")
+  │
+  └ 1   x + y >= x
+        Return: x + y
+```
+
+Each leaf on the tree represents a possible execution path, and has some logical conditions
+attached:
+
+- `0`:  `msg.value > 0`: revert (`add` is not payable)
+- `1-0`: `msg.value == 0 && x + y < x`: revert (overflow)
+- `1-1`: `msg.value == 0 && x + y >= x`: return `x + y`
+
+If we assert a property at every leaf in the execution tree, then we can be sure that that property
+will hold for all possible values of each piece of symbolic state, allowing us to prove properties
+that hold for *all possible inputs* to a given function.
+
+## Using `ds-test`
+
+[`ds-test`](https://github.com/dapphub/ds-test/blob/master/src/test.sol) is a solidity unit testing
```

Comment:
Solidity is written as `Solidity` earlier and `solidity` here.

Reply:
which do you prefer?

Reply:
I personally prefer Solidity